include/seqan3/io/alignment_file/format_bam.hpp

```diff
@@ -0,0 +1,583 @@
+
+            int32_t n_ref;
+            read_field(stream_view, n_ref);
+
+            for (int32_t ref_idx = 0; ref_idx < n_ref; ++ref_idx)
+            {
+                read_field(stream_view, tmp32);
+                std::string id; // what if no string?
+                id.resize(tmp32 - 1);
+                std::ranges::copy_n(std::ranges::begin(stream_view), tmp32 - 1, &id[0]); // copy without \0 character
+                ++std::ranges::begin(stream_view); // skip \0 character
+                read_field(stream_view, tmp32);
+
+                auto id_it = header.ref_dict.find(id);
+
+                if (id_it == header.ref_dict.end())
+                    throw format_error{"Unknown reference name " + id + " found in BAM header."};
+
+                if (id_it->second != ref_idx)
+                    throw format_error{detail::to_string("Reference id '", id, "' at position ", ref_idx,
+                                                         " does not correspond to the position ", id_it->second,
+                                                         " in the header (header.ref_ids():", header.ref_ids(), ").")};
+
+                if (std::get<0>(header.ref_id_info[id_it->second]) != tmp32)
+                    throw format_error{"Provided reference has unequal length as specified in the header."};
+            }
+
+            header_was_read = true;
+
+            if (stream_buf_t{stream} == stream_buf_t{}) // no records follow
+                return;
+        }
+
+        // read alignment record into buffer
+        // -------------------------------------------------------------------------------------------------------------
+        alignment_record_core record;
+        std::ranges::copy_n(stream_view.begin(), sizeof(record), reinterpret_cast<char*>(&record));
+
+        if (record.refID >= static_cast<int32_t>(header.ref_ids().size()) || record.refID < -1) // [[unlikely]]
+            throw format_error{"Reference id index '" + std::to_string(record.refID) +
+                               "' is not in range of header.ref_ids()."};
+
+        if (record.refID > -1) // not unmapped
+            ref_id = record.refID;                                                   // field::REF_ID
+        flag = record.flag;                                                          // field::FLAG
+        mapq = record.mapq;                                                          // field::MAPQ
+
+        if (record.pos > -1) // [[likely]]
+            ref_offset = record.pos;                                                 // field::REF_OFFSET
+
+        if constexpr (!detail::decays_to_ignore_v<mate_type>)                        // field::MATE
+        {
+            if (record.next_refID > -1)
+                get<0>(mate) = record.next_refID;
+
+            if (record.next_pos > -1) // [[likely]]
+                get<1>(mate) = record.next_pos;
+
+            get<2>(mate) = record.tlen;
```

Comment:
I not sure to what you refer:
yes, the mate is set since I check `if constexpr (!detail::decays_to_ignore_v<mate_type>)`
yes, the `record.tlen` always filled/set because the BAM spec have this as a fixed parameter of the record that I read in a batch above.

Reply:
Yes, I meant the tlen parameter.